ops_tests/test_upgrade_charm.py

```diff
@@ -0,0 +1,78 @@
+import logging
+import subprocess
+import tempfile
+import asyncio
+
+from juju.errors import JujuError
+import yaml
+
+import pytest
+
+log = logging.getLogger(__name__)
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_old_bundle(ops_test):
+    await ops_test.model.deploy("cs:kubeflow")
+    subprocess.run(
+        [
+            'juju',
+            'kubectl',
+            'patch',
+            'role/istio-ingressgateway-operator',
+            '-p',
+            yaml.dump(
+                {
+                    "apiVersion": "rbac.authorization.k8s.io/v1",
+                    "kind": "Role",
+                    "metadata": {"name": "istio-ingressgateway-operator"},
+                    "rules": [{"apiGroups": ["*"], "resources": ["*"], "verbs": ["*"]}],
+                }
+            ),
+        ]
+    )
+    # raise_on_error=False due to kubeflow-profiles going to an error state if
+    # kubeflow-dashboard CRDs aren't created yet.
+    await ops_test.model.wait_for_idle(raise_on_error=False)
+
+
+async def test_upgrade_apps_in_bundle(ops_test):
+    res = subprocess.run(["charm", "-v"])
+    if res.returncode != 0:
+        logging.error(
+            "Charm snap is not installed, please install with: \n \
+                sudo snap install charm --classic"
+        )
+
+    temp_dir = f"{tempfile.mkdtemp()}/kubeflow-bundle"
```

Comment:
This seems like it will leave old directories around. Can we run it like this?

```python
with tempfile.TemporaryDirectory() as temp_dir:
    ...
```